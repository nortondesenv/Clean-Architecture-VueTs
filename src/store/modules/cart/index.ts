import { Module } from 'vuex';
import { RootState } from '@/store';
import { mutations } from './mutations';
import { CartState, initialState } from './state';

export * from './state';

export const store: Module<CartState, RootState> = {
    state: initialState,
    mutations,
};
