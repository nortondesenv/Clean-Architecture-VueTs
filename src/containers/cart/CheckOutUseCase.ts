import BaseUseCase from '@/usecases/BaseUseCase';
import ErrorService from '@/services/ErrorService';
import { CartEntity } from '@/entities/Cart';

export interface ICheckOutUseCase {
    cart: CartEntity;
    errorService: ErrorService;
}

export default class CheckOutUseCase implements BaseUseCase {
    cart: CartEntity;
    errorService: ErrorService;

    constructor({ cart, errorService }: ICheckOutUseCase) {
        this.cart = cart;
        this.errorService = errorService;
    }

    async execute() {
        try {
            await this.cart.checkOut();
            alert('Obrigado e volte sempre!');
        } catch (e) {
            await this.errorService.handle(e);
        }
        return;
    }
}
