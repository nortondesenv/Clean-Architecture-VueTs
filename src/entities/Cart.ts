import { IProduct } from '@/entities/Product';
import store from '@/store';
import { AddItem, ClearItems, ToggleCart } from '@/store/modules/cart/types';

export interface ICartItem extends IProduct {
    checkoutPrice: number;
    amount: number;
}

export interface ICart {
    items: ICartItem[];
}

export class CartEntity {
    constructor() {}

    async toggleCart() {
        await store.commit(new ToggleCart(store.state.cart.isCartOpen));
    }

    removeItem(index: number) {
        store.state.cart.cart.items.splice(index, 1);
    }

    async addItemToCart(item: IProduct) {
        const checkoutPrice = item.priceMember;
        const amount = 1;
        const cartItem = {
            ...item,
            checkoutPrice,
            amount,
        };
        await store.commit(new AddItem(cartItem));
        alert('Produto adicionado em seu carrinho.');
    }

    async checkOut() {
        // Start operation
        store.commit(new ClearItems());
        alert('Obrigado e volte sempre!');
        return;
    }
}
