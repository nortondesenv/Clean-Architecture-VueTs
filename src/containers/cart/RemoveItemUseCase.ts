import BaseUseCase from '@/usecases/BaseUseCase';
import ErrorService from '@/services/ErrorService';
import { CartEntity } from '@/entities/Cart';

export interface IRemoveItemUseCase {
    cart: CartEntity;
    errorService: ErrorService;
}

export default class RemoveItemUseCase implements BaseUseCase {
    cart: CartEntity;
    errorService: ErrorService;

    constructor({ cart, errorService }: IRemoveItemUseCase) {
        this.cart = cart;
        this.errorService = errorService;
    }

    async execute(params) {
        try {
            await this.cart.removeItem(params);
        } catch (e) {
            await this.errorService.handle(e);
        }
        return;
    }
}
