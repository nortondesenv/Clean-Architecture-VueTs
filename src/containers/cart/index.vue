<template>
    <transition name="fade">
        <div v-if="presenter.isCartOpen">
            <div class="Cart">
                <template>
                    <div class="Cart__Divisor--Top">
                        <CartHeader
                            @toggle-cart="toggleCart()"
                            :cartCount="cartCount"
                        ></CartHeader>
                    </div>
                    <div class="Cart__Divisor--Mid">
                        <CartEmpty v-if="presenter.isEmpty"></CartEmpty>
                        <CartProduct
                            v-else
                            :cartOpen="true"
                            v-for="(item, idx) in presenter.items"
                            :key="idx"
                            :item="item"
                            :index="idx"
                        />
                    </div>
                    <div class="Cart__Divisor--Bot">
                        <CartCheckOut :totalPrice="totalPrice"></CartCheckOut>
                    </div>
                </template>
            </div>
            <div @click="toggleCart()" class="Cart__Overlay"></div>
        </div>
    </transition>
</template>

<script lang="ts">
import Vue from 'vue';

import Presenter, { IPresenter, IPresenterState } from './presenter';
import ToggleCartUseCase, {
    IToggleCartUseCase,
} from '@/usecases/ToggleCartUseCase';

import { CartEntity } from '@/entities/Cart';
import CartRepository from '@/repositories/CartRepository';

// Components
import CartProduct from '@/components/Modules/CartProduct.vue';
import CartHeader from '@/components/Modules/CartHeader.vue';
import CartCheckOut from '@/components/Modules/CartCheckOut.vue';
import CartEmpty from '@/components/Modules/CartEmpty.vue';

export default Vue.extend({
    components: {
        CartProduct,
        CartHeader,
        CartEmpty,
        CartCheckOut,
    },
    props: {
        totalPrice: {
            type: Number,
            default: 0,
        },
        cartCount: {
            type: Number,
            default: 0,
        },
    },
    computed: {
        presenter(): IPresenterState {
            const params: IPresenter = {
                cartRepository: new CartRepository(),
            };
            return Presenter(params);
        },
    },
    methods: {
        async toggleCart() {
            if (!this.presenter.isCartOpen) return;

            const params: IToggleCartUseCase = {
                cart: new CartEntity(),
            };
            await new ToggleCartUseCase(params).execute();
        },
    },
});
</script>
