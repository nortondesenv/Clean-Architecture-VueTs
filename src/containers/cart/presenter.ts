import { ICartItem } from '@/entities/Cart';
import CartRepository from '@/repositories/CartRepository';

export interface IPresenter {
    cartRepository: CartRepository;
}

export interface IPresenterState {
    items: ICartItem[];
    isEmpty: boolean;
    isCartOpen: boolean;
}

export default ({ cartRepository }: IPresenter): IPresenterState => {
    const items = cartRepository.getItems();
    return {
        items,
        isEmpty: !items || items.length == 0,
        isCartOpen: cartRepository.isCartOpen(),
    };
};
