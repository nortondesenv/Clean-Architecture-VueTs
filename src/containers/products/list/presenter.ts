import { IProduct } from '@/entities/Product';
import ProductRepository from '@/repositories/ProductRepository';

export interface IPresenter {
    productRepository: ProductRepository;
}

export interface IPresenterState {
    items: List<IProduct>;
}

export default ({ productRepository }: IPresenter): IPresenterState => {
    return {
        items: productRepository.getItems(),
    };
};
