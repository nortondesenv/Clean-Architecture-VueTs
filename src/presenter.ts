import CartRepository from '@/repositories/CartRepository';
import { ICartItem } from '@/entities/Cart';

export interface IPresenter {
    cartRepository: CartRepository;
}

export interface IPresenterState {
    totalPrice: number;
    cartItems: ICartItem[];
    cartCount: number;
}

const calcTotalPrice = (cartRepository: CartRepository): number => {
    const cartItems = cartRepository.getItems();
    const totalPrice = cartItems.reduce((prev, current) => {
        return prev + current.checkoutPrice * current.amount;
    }, 0);
    return totalPrice;
};

export default ({ cartRepository }: IPresenter): IPresenterState => {
    return {
        totalPrice: calcTotalPrice(cartRepository),
        cartItems: cartRepository.getItems(),
        cartCount: cartRepository.countCart(),
    };
};
