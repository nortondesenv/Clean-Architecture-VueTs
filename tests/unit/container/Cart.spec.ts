import { shallowMount } from '@vue/test-utils';
import Cart from '@/containers/cart/index.vue';
import { IPresenterState } from '@/containers/cart/presenter';
import { ICartItem } from '@/entities/Cart';

describe('Header.vue', () => {
    const cartItem: ICartItem[] = [];

    const presenterCart: IPresenterState = {
        items: cartItem,
        isEmpty: false,
        isCartOpen: false,
    };

    it('Cart propriedade fechada', () => {
        const wrapper = shallowMount(Cart, {
            computed: {
                presenter() {
                    return presenterCart;
                },
            },
        });
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Cart propriedade aberta', () => {
        presenterCart.isCartOpen = true;
        const wrapper = shallowMount(Cart, {
            computed: {
                presenter() {
                    return presenterCart;
                },
            },
        });
        expect(wrapper.html()).toContain('Cart__Divisor--Top');
    });

    it('Cart vazio', () => {
        presenterCart.isEmpty = true;
        const wrapper = shallowMount(Cart, {
            computed: {
                presenter() {
                    return presenterCart;
                },
            },
        });
        expect(wrapper.html()).toContain('cartempty');
    });
});
