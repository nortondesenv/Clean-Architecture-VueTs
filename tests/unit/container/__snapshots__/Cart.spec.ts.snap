// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.vue Cart propriedade fechada 1`] = `
<transition-stub name="fade">
  <!---->
</transition-stub>
`;
